### PR TITLE
Add flexible journal vdev.

### PR DESCRIPTION
**Pull request by sanebay**
Ability to add chunks to a journal vdev dynamically. Create pool of chunks and maintain a list of chunks sliding window. Add api to create chunk. Chunks part of journal vdev are connected like singly linked list with next_chunk in private data. Chunks part of journal vdv have private data. Journal vdev can give independent descriptors to the user to do alloc write and truncate. Modified tests to support this.

Changes made.
1. Earlier only 2 journalvdev were there (data and control) and took fixed size and numchunks. Now user can only provide chunk size and we add chunks dynamically once vdev is full. Most of the state of the list of chunks and its offsets were maintained in the JournalVDev class now its moved to its inner class JournalVDev::Descriptor. Now a single JournalVDev can create multiple descriptors. Each descriptor now maintains the list of chunks, its offsets etc. User call open to create different descriptors.
2. Earlier the offsets of the sliding window could wrap around due to the fixed number of chunks. But now they only increase. Both the left and right side of the window increase. Truncate will increase the left and append blk will increase the right offset.
3. Add a chunk pool in journal vdev so that its faster to append a new chunk to the list.
4. Changes in device_manager is to support to start with zero chunks and api to create and remove single chunk.
5. Test code is almost same. Earlier the api were tested against a JournalVirtualDev but now its against a JournalVirtualDev::Descriptor. Added recovery test case.
6. JournalVDev ......................   Passed   38.81 sec

### CODE REVIEW

File: src/lib/common/homestore_config.fbs [hunk 1]
```diff
@@ -140,6 +140,9 @@ table Generic {
     // percentage of cache used to create indx mempool. It should be more than 100 to 
     // take into account some floating buffers in writeback cache.
     indx_mempool_percent : uint32 = 110;
+
+    // journal chunk pool size
+    journal_chunk_pool_size: uint32 = 5;
```
Comment (hkadayam) [inline src/lib/common/homestore_config.fbs:145]: can you please add unit as well, something like journal_chunk_pool_size_mb

File: src/lib/device/device.h [hunk 2]
```diff
@@ -95,10 +96,11 @@ ENUM(chunk_selector_t, uint8_t, // What are the options to select chunk to alloc
 struct vdev_parameters {
     std::string vdev_name;                  // Name of the vdev
     uint64_t vdev_size;                     // Current Vdev size.
-    uint32_t num_chunks;                    // Total number of primary chunks.
+    uint32_t num_chunks{};                  // Total number of primary chunks.
                                             // NOTE: If pdev opts is ALL_PDEV_STRIPED, then num_chunks would round off
                                             // to number of pdevs evenly
     uint32_t blk_size;                      // Block size vdev operates on
+    uint32_t chunk_size{};                  // Chunk size provided for dynamic vdev.
```
Comment (hkadayam) [inline src/lib/device/device.h:103]: If user has to define either chunk_size or num_chunks, should we put them as std::variant?

File: src/lib/device/virtual_dev.hpp [hunk 3]
```diff
@@ -92,7 +92,7 @@ class VirtualDev {
 
     std::mutex m_mgmt_mutex;          // Any mutex taken for management operations (like adding/removing chunks).
     std::set< PhysicalDev* > m_pdevs; // PDevs this vdev is working on
-    sisl::sparse_vector< shared< Chunk > > m_all_chunks; // All chunks part of this vdev
+    std::vector< shared< Chunk > > m_all_chunks; // All chunks part of this vdev
```
Comment (hkadayam) [inline src/lib/device/virtual_dev.hpp:95]: Why is it replaced with std::vector<>? Chunks could be read in any order right and we wanted to place them in chunk_id order right?

File: src/lib/logstore/log_stream.cpp [hunk 4]
```diff
@@ -22,13 +22,16 @@
...
 namespace homestore {
 SISL_LOGGING_DECL(logstore)
 
-log_stream_reader::log_stream_reader(off_t device_cursor, JournalVirtualDev* store, uint64_t read_size_multiple) :
-        m_vdev{store}, m_first_group_cursor{device_cursor}, m_read_size_multiple{read_size_multiple} {
-    m_vdev->lseek(m_first_group_cursor);
+log_stream_reader::log_stream_reader(off_t device_cursor, JournalVirtualDev* vdev,
+                                     shared< JournalVirtualDev::Descriptor > vdev_jd, uint64_t read_size_multiple) :
+        m_vdev{vdev},
+        m_vdev_jd{vdev_jd},
```
Comment (hkadayam) [inline src/lib/logstore/log_stream.cpp:31]: nit: Can do ```m_vdev_js{std::move(vdev_jd)}```

File: src/lib/common/homestore_config.fbs (hunk 1 above)
Comment (sanebay) [reply src/lib/common/homestore_config.fbs:145]: @hari, add num_

File: src/lib/device/journal_vdev.hpp [hunk 5]
```diff
@@ -271,66 +397,29 @@ class JournalVirtualDev : public VirtualDev {
...
+    // Chunk pool is used to get chunks when there is no space
+    // and its cheaper compared to create a chunk on the fly.
+    // Creating chunk on the fly causes sync write. Getting a chunk
+    // from the pool causes a single write of the last chunk in the list
+    // to update its end_of_chunk and next_chunk.
+    std::list< shared< Chunk > > m_chunk_pool;
+    uint32_t m_chunk_pool_capacity;
+    std::condition_variable m_chunk_pool_cv;
+    std::mutex m_chunk_pool_mutex;
+    iomgr::io_fiber_t m_chunk_pool_fiber;
```
Comment (sanebay) [inline src/lib/device/journal_vdev.hpp:420]: hari: use normal std thread

File: src/lib/device/journal_vdev.cpp [hunk 6]
```diff
@@ -34,75 +34,300 @@
...
+        uint64_t created_at{};
+    };
+
+    std::unordered_map< logdev_id_t, HeadChunk > logdev_head_map;
+    std::unordered_map< chunk_num_t, shared< Chunk > > chunk_map;
+    std::unordered_set< chunk_num_t > visited_chunks;
+
+    // Traverse the chunks and find the heads of the logdev_id's.
+    for (auto& chunk : m_all_chunks) {
+        auto* data = r_cast< JournalChunkPrivate* >(const_cast< uint8_t* >(chunk->user_private()));
```
Comment (sanebay) [inline src/lib/device/journal_vdev.cpp:64]: descriptor_chunk_info

File: src/lib/device/journal_vdev.cpp [hunk 7]
```diff
@@ -34,75 +34,300 @@
...
+    m_chunk_pool_cv.notify_one();
+    return chunk;
+}
+
+void JournalVirtualDev::release_chunk_to_pool(const shared< Chunk >& chunk) {
+    RELEASE_ASSERT(chunk, "Chunk invalid");
+    bool reuse = false;
+    {
+        std::unique_lock< std::mutex > lk{m_chunk_pool_mutex};
+        if (m_chunk_pool.size() < m_chunk_pool_capacity) {
```
Comment (sanebay) [inline src/lib/device/journal_vdev.cpp:209]: hari: clear the private data before putting back to pool

File: src/lib/device/journal_vdev.cpp [hunk 8]
```diff
@@ -34,75 +34,300 @@
...
+            m_chunk_pool.push_back(chunk);
+            HS_LOG(TRACE, device, "Produced chunk to pool id {}", chunk->chunk_id());
+        }
+    });
+}
+
+shared< Chunk > JournalVirtualDev::get_chunk_from_pool() {
+    shared< Chunk > chunk;
+    {
+        std::unique_lock< std::mutex > lk{m_chunk_pool_mutex};
```
Comment (sanebay) [inline src/lib/device/journal_vdev.cpp:194]: wait if the queue is empty.

File: src/lib/common/homestore_config.fbs (hunk 1 above)
Comment (sanebay) [reply src/lib/common/homestore_config.fbs:145]: Ack

File: src/lib/device/virtual_dev.hpp (hunk 3 above)
Comment (sanebay) [reply src/lib/device/virtual_dev.hpp:95]: Reverted.

File: src/lib/logstore/log_stream.cpp (hunk 4 above)
Comment (sanebay) [reply src/lib/logstore/log_stream.cpp:31]: Ack

File: src/lib/device/journal_vdev.hpp (hunk 5 above)
Comment (sanebay) [reply src/lib/device/journal_vdev.hpp:420]: Moving the chunk_pool to chunk.cpp

File: src/lib/device/device.h (hunk 2 above)
Comment (sanebay) [reply src/lib/device/device.h:103]: As discussed added vdev_size_type_t enum {STATIC, DYNAMIC} in format param. If it static it takes size and num_chunks or chunk_size. If its dynamic,  it needs only chunk_size.

File: src/lib/device/device.h [hunk 9]
```diff
@@ -181,4 +189,52 @@ class DeviceManager {
...
+    // Returns the capacity of the chunk pool.
+    uint64_t capacity() { return m_params.pool_capacity; }
+    uint64_t size() { return m_pool.size(); }
+
+private:
+    // Producer thread.
+    void producer();
+
+private:
+    // Chunk pool is used to get chunks when there is no space
```
Comment (yamingk) [inline src/lib/device/device.h:226]: move this comment to head of the class as some documentation of this class.

File: src/lib/device/journal_vdev.hpp [hunk 10]
```diff
@@ -19,250 +19,367 @@
...
+    // right side (m_end_offset). Truncate increases the left m_data_start_offset
+    // and pop chunks from the front of the list. alloc_next_append_blk adds more chunk to
+    // the back of list if no space and adjusts the right m_end_offset. All offsets
+    // only increase and never wraps around. Each chunk in the descriptor has private
+    // data about the logdev_id its part of, next chunk to maintain the list.
+    struct Descriptor {
+    private:
+        JournalVirtualDev& m_vdev;
+        logdev_id_t m_logdev_id; // Unique id identifying the journal descriptor.
+        // off_t is long. make it uint64_t ?
```
Comment (yamingk) [inline src/lib/device/journal_vdev.hpp:62]: good point, long is only 4 bytes, and it is signed, it will overflow if exceeding 2TB offset, right?

File: src/lib/device/journal_vdev.cpp [hunk 11]
```diff
@@ -34,75 +34,211 @@
...
+    RELEASE_ASSERT_LT(sz, chunk_size, "Size requested greater than chunk size");
+
+    if ((tail_offset() + static_cast< off_t >(sz)) >= m_end_offset) {
+        // not enough space left, add a new chunk.
+        HS_LOG(DEBUG, device,
+               "No space left, desc: {} m_write_sz_in_total: {}, m_reserved_sz: {} tail: 0x{}. Creating chunk",
+               m_logdev_id, m_write_sz_in_total.load(), m_reserved_sz, to_hex(tail_offset()));
+
+#ifdef _PRERELEASE
+        iomgr_flip::test_and_abort("abort_before_update_eof_cur_chunk");
```
Comment (yamingk) [inline src/lib/device/journal_vdev.cpp:212]: Are these flips enabled and running fine somewhere?

File: src/lib/device/journal_vdev.hpp [hunk 12]
```diff
@@ -19,250 +19,367 @@
...
+using logdev_id_t = uint64_t;
+
+// Chunks used for journal vdev has journal related info stored in chunk private data.
+// Each log device has a list of journal chunk data with next_chunk.
+// Journal vdev will arrange the chunks in order during recovery.
+struct JournalChunkPrivate {
+    logdev_id_t logdev_id{0};
+    bool is_head{false};           // Is it the head element.
+    uint64_t created_at{0};        // Creation timestamp
+    uint64_t end_of_chunk_size{0}; // The offset indicates end of chunk.
```
Comment (yamingk) [inline src/lib/device/journal_vdev.hpp:42]: nit: I know the this is migrated from old, can we rename it to 'end_offset', since it is actually a offset not a size.

File: src/lib/device/virtual_dev.cpp [hunk 13]
```diff
@@ -123,6 +123,14 @@ void VirtualDev::add_chunk(cshared< Chunk >& chunk, bool is_fresh_chunk) {
     m_chunk_selector->add_chunk(chunk);
 }
 
+void VirtualDev::remove_chunk(cshared< Chunk >& chunk) {
+    std::unique_lock lg{m_mgmt_mutex};
+    auto iter = std::remove_if(m_all_chunks.begin(), m_all_chunks.end(), [chunk](auto c) { return c == chunk; });
+    m_all_chunks.erase(iter, m_all_chunks.end());
+
+    // m_chunk_selector->remove_chunk(chunk); // TODO
```
Comment (yamingk) [inline src/lib/device/virtual_dev.cpp:131]: won't there be problem if it is not removed from heap_chunk_selector, is there a reason it can't be done within this PR?

File: src/lib/device/journal_vdev.cpp [hunk 14]
```diff
@@ -34,75 +34,211 @@
...
+        return journal_desc;
+    }
+
+    HS_LOG(TRACE, device, "Opened log device descriptor {}", logdev_id);
+    return it->second;
+}
+
+void JournalVirtualDev::Descriptor::append_chunk() {
+    // Get a new chunk from the pool.
+    auto new_chunk = m_vdev.m_chunk_pool->dequeue();
```
Comment (yamingk) [inline src/lib/device/journal_vdev.cpp:160]: will there be a case when m_chunk_pool is empty and new_chunk is null?

File: src/lib/device/journal_vdev.hpp [hunk 15]
```diff
@@ -19,250 +19,367 @@
...
+        off_t dev_offset(off_t nbytes) const;
+
+        /**
+         * @brief : get the start logical offset where data starts;
+         *
+         * @return : the start logical offset where data starts;
+         */
+        off_t data_start_offset() const { return m_data_start_offset; }
+
+        off_t end_offset() const { return m_end_offset; }
```
Comment (yamingk) [inline src/lib/device/journal_vdev.hpp:228]: can you also provide comment of this API, especially the difference between tail_offset, so that who are not familiar with this code can be easier to understand.

File: src/lib/device/journal_vdev.cpp [hunk 16]
```diff
@@ -34,75 +34,211 @@
...
-    if (end_offset < ds_off) { HS_DBG_ASSERT_EQ(size() - used_size(), static_cast< uint64_t >(ds_off - end_offset)); }
-#endif
-    // works for both "end_offset >= ds_off" and "end_offset < ds_off";
-    if (offset_in_chunk + sz <= chunk->size()) {
-        // not acrossing boundary, nothing to do;
-    } else if ((used_size() + (chunk->size() - offset_in_chunk) + sz) <= size()) {
-        // across chunk boundary, still enough space;
+        // Clear the private chunk data.
+        *data = JournalChunkPrivate{};
+        update_chunk_private(chunk, data);
```
Comment (yamingk) [inline src/lib/device/journal_vdev.cpp:118]: shouldn't we set the end_offset in private data here, e.g. like 69 was doing?

File: src/lib/device/journal_vdev.cpp [hunk 17]
```diff
@@ -34,75 +34,211 @@
...
+            m_chunk_pool->enqueue(chunk);
+            LOGWARN("Reuse to pool orphan chunk {} found for logdev {} next {}.", chunk_id, logdev_id, next_chunk);
+        } else {
+            LOGWARN("Removing orphan chunk {} found for logdev {} next {}.", chunk_id, logdev_id, next_chunk);
+            m_dmgr.remove_chunk(chunk);
+        }
+    }
 
-        // Update the overhead to total write size;
-        m_write_sz_in_total.fetch_add(chunk->size() - offset_in_chunk, std::memory_order_relaxed);
```
Comment (yamingk) [inline src/lib/device/journal_vdev.cpp:66]: this line should be kept, no?

File: src/lib/device/journal_vdev.cpp [hunk 18]
```diff
@@ -34,75 +34,211 @@
...
+        auto last_chunk = m_journal_chunks.back();
+        auto* last_chunk_private = r_cast< JournalChunkPrivate* >(const_cast< uint8_t* >(last_chunk->user_private()));
+
+        // Set the next chunk with the newly created chunk id.
+        last_chunk_private->next_chunk = new_chunk->chunk_id();
+
+        // Append the new chunk
+        m_journal_chunks.push_back(new_chunk);
+        auto chunk_size = m_vdev.info().chunk_size;
+        auto offset_in_chunk = (tail_offset() % chunk_size);
```
Comment (yamingk) [inline src/lib/device/journal_vdev.cpp:179]: suggest to call offset_to_chunk to not assume chunk is always same size.

File: src/lib/device/virtual_dev.cpp (hunk 13 above)
Comment (sanebay) [reply src/lib/device/virtual_dev.cpp:131]: Added remove_chunk() default function. HO can override it.

File: src/lib/device/journal_vdev.cpp (hunk 11 above)
Comment (sanebay) [reply src/lib/device/journal_vdev.cpp:212]: We need to add it in our long running tests

File: src/lib/device/journal_vdev.hpp (hunk 12 above)
Comment (sanebay) [reply src/lib/device/journal_vdev.hpp:42]: Made it end_of_chunk

File: src/lib/device/journal_vdev.hpp (hunk 10 above)
Comment (sanebay) [reply src/lib/device/journal_vdev.hpp:62]: Currently off_t is mapped to int64_t which itself is huge.

File: src/lib/device/journal_vdev.hpp (hunk 12 above)
Comment (sanebay) [reply src/lib/device/journal_vdev.hpp:42]: Made it end_of_chunk

File: src/lib/device/journal_vdev.cpp (hunk 14 above)
Comment (sanebay) [reply src/lib/device/journal_vdev.cpp:160]: Added assert.

File: src/lib/device/journal_vdev.hpp (hunk 15 above)
Comment (sanebay) [reply src/lib/device/journal_vdev.hpp:228]: Its added near the class variables. journal_vdev.hpp:50. If its not clear, please let me know.

File: src/lib/device/journal_vdev.cpp (hunk 18 above)
Comment (sanebay) [reply src/lib/device/journal_vdev.cpp:179]: As discussed all chunks in a vdev have same size.

File: src/lib/device/journal_vdev.cpp (hunk 17 above)
Comment (sanebay) [reply src/lib/device/journal_vdev.cpp:66]: This is now done in append_chunk() function line 183

File: src/tests/test_journal_vdev.cpp [hunk 19]
```diff
@@ -315,20 +360,105 @@ class VDevIOTest : public ::testing::Test {
...
-    uint64_t m_total_size = 0;
     std::map< off_t, write_info > m_off_to_info_map;
     Clock::time_point m_start_time;
-    std::shared_ptr< JournalVirtualDev > m_vdev;
+    std::shared_ptr< JournalVirtualDev::Descriptor > m_vdev_jd;
+    friend class VDevJournalIOTest;
 };
 
-TEST_F(VDevIOTest, VDevIOTest) { this->execute(); }
+// TODO add more tests covering unclean shutdown and more corner cases.
```
Comment (yamingk) [inline src/tests/test_journal_vdev.cpp:375]: Is there a issue created for this?

File: src/tests/test_journal_vdev.cpp [hunk 20]
```diff
@@ -315,20 +360,105 @@ class VDevIOTest : public ::testing::Test {
...
+
+TEST_F(VDevJournalIOTest, LongRunning) {
```
Comment (yamingk) [inline src/tests/test_journal_vdev.cpp:377]: Can we also specify how long has it been run in the PR description, would be helpful for reviewers?

It should be good to also add this long running test case to existing logstore long running, right?

File: src/tests/test_journal_vdev.cpp (hunk 19 above)
Comment (sanebay) [reply src/tests/test_journal_vdev.cpp:375]: New issue created. Will add that as future task https://github.com/eBay/HomeStore/issues/283

File: src/tests/test_journal_vdev.cpp (hunk 20 above)
Comment (sanebay) [reply src/tests/test_journal_vdev.cpp:377]: Currently its running just 40secs. Yes we should add it to long running.

**Review by yamingk — APPROVED** [review]
